sdk/servicebus/Azure.Messaging.ServiceBus/MigrationGuide.md

```diff
@@ -42,210 +38,221 @@ In the case of Service Bus, the modern client libraries have packages and namesp
 
-### Sending and receiving a batch of messages
+// define the message handler
+async Task MessageHandler(Message message, CancellationToken token)
+{
+    // Process the message.
+    Console.WriteLine($"Received message with Body:{Encoding.UTF8.GetString(message.Body)}");
+}
 
-In v4, `QueueClient`/`MessageSender`/`MessageReceiver` would be created directly, after which user would call `SendAsync()` method via `QueueClient`/`MessageSender` to send a batch of messages and `ReceiveAsync()` method via `MessageReceiver` to receive a batch of messages.
+// define the error handler
+Task ErrorHandler(ExceptionReceivedEventArgs exceptionReceivedEventArgs)
+{
+    Console.WriteLine($"Message handler encountered an exception {exceptionReceivedEventArgs.Exception}.");
+    return Task.CompletedTask;
+}
 
-In v7, user would initialize the `ServiceBusClient` and call `CreateSender()` method to create a `ServiceBusSender` and `CreateReceiver()` method to create a `ServiceBusReceiver`. There are two ways of sending several messages at once.
+// Configure the message handler options with the error handler
+var messageHandlerOptions = new MessageHandlerOptions(ErrorHandler);
 
-The first way uses the `SendMessagesAsync`overload that accepts an IEnumerable of `ServiceBusMessage`. With this method, we will attempt to fit all of the supplied messages in a single message batch that we will send to the service. If the messages are too large to fit in a single batch, the operation will throw an exception.
+// Register the message handler and options
+queueClient.RegisterMessageHandler(MessageHandler, messageHandlerOptions);
 
-The second way of doing this is using safe-batching. With safe-batching, you can create a `ServiceBusMessageBatch` object, which will allow you to attempt to add messages one at a time to the batch using the `TryAddMessage` method. If the message cannot fit in the batch, `TryAddMessage` will return false. If the `ServiceBusMessageBatch` accepts a message, user can be confident that it will not violate size constraints when calling `SendMessagesAsync()` via `ServiceBusSender`. To receive a set of messages, a user would call `ReceiveMessagesAsync()` method via `ServiceBusReceiver`.
+// Or receive using the receiver
+var receiver = new MessageReceiver(connectionString, queueName);
+var receivedMessage = await receiver.ReceiveAsync();
+Console.WriteLine($"Received message with Body:{Encoding.UTF8.GetString(receivedMessage.Body)}");
+await receiver.CompleteAsync(receivedMessage);
+```
 
-In v4:
+Now in `Azure.Messaging.ServiceBus`, we introduce the concept of a "processor" which takes your message and error handlers to provide you a simple way to get started with processing your messages. This provides a graceful shutdown via the `StopProcessingAsync` method which will ensure that no more messages will be received, but at the same time you can continue the prcoessing and settling the messages already in flight.
```

Comment:
That's inaccurate. This option has existed both in track 0 and track 1.
What has been done are some improvements to make it more robust (stopping the handler/processor).

Reply:
Are you referring to the message and error handlers being available in track 0 and 1, or the shutdown that allows one to continue processing and settling messages in flight, but stop receiving more messages?

Reply:
The earlier. Specifically, this sentence:
> Now in `Azure.Messaging.ServiceBus`, we introduce the concept of a "processor" which takes your message and error handlers to provide you a simple way to get started with processing your messages.

The graceful shutdown is the improvement I've mentioned in my 2nd sentence.
> What has been done are some improvements to make it more robust (stopping the handler/processor).

Reply:
Understood. What I was going for was to note that there is a dedicated processor class for the said feature. How about the below?

```suggestion
Now in `Azure.Messaging.ServiceBus`, we introduce a dedicated class `ServiceBusProcessor` which takes your message and error handlers to provide you the same simple way to get started with processing your messages with auto complete and auto lock renewal features. This also provides a graceful shutdown via the `StopProcessingAsync` method which will ensure that no more messages will be received, but at the same time you can continue the prcoessing and settling the messages already in flight.
```

Reply:
Almost. When you say

> ...to provide you the same simple way to get started with processing your messages with auto complete and auto lock renewal features. 

you're missing something. "the same simple way" compared to what? 😀

How about this?

```suggestion
Now in `Azure.Messaging.ServiceBus`, we introduce a dedicated class `ServiceBusProcessor` which takes your message and error handlers to provide you with the same simple way to get started with processing your messages as message handlers in the previous packages, with auto-complete and auto-lock renewal features. This class also provides a graceful shutdown via the `StopProcessingAsync` method which will ensure that no more messages will be received, but at the same time you can continue the processing and settling the messages already in flight.
```